### ISSUE TITLE
Add 'bazel debug'

### ISSUE THREAD

**Issue by kayasoze** [bazelbuild/bazel#312]
A `bazel debug` command would be nice to have for local binaries.

As a workaround, `bazel run --config=debug` could likely be defined in .bazelrc to use something like `--run_under=<debug wrapper>`.

**Comment by cgrushko** [comment]
Doesn't solve everything, but there's `bazel tests --java_debug` which waits for a Java debugger to attach before running the test.

    --java_debug
      Causes the Java virtual machine of a java test to wait for a connection
      from a JDWP-compliant debugger (such as jdb) before starting the test.
      Implies -test_output=streamed.
      Expands to: --test_arg=--wrapper_script_flag=--debug --test_output=streamed
      --test_strategy=exclusive --test_timeout=9999 --nocache_test_results

**Comment by laszlocsomor** [comment]
Please reopen if Carmi's suggested solution is insufficient.

**Comment by laszlocsomor** [comment]
I mean, please ping this bug.
(I just learned that only team members can reopen bugs.)

**Comment by cgrushko** [comment]
By the way - I don't think Bazel has something like that for `blaze run`, which I find annoying sometimes (starting a Java binary under a debugger requires an incantation I never remember).

Also, we don't have something like this for C++ and Python.

**Comment by kayasoze** [comment]
My primary languages are C++ and Python.

**Comment by kayasoze** [comment]
Please reopen for C++ and Python.

**Comment by davidzchen** [comment]
I think it would be good to have something that can be generic for other languages and not just Java.

Reopening for feedback.

**Comment by ulfjack** [comment]
Well, someone needs to come up with a proposal of how to do it and then we can iterate on that. --java_debug is special; it uses a feature of the wrapper script we generate for java_binary and java_test rules (which adds the --debug flag), but it's not clear how that could work for C++ binaries (for example), since there's no part that's under Bazel's control. Python could work similar to Java - except I have no idea how you'd even go about debuggin a python binary...

**Comment by kayasoze** [comment]
C++ debugging *almost* works with `bazel run --compilation_mode=dbg --run_under=/usr/bin/lldb` (at least with lldb), however, stdin would need to be left open, and currently it is closed.

**Comment by ulfjack** [comment]
Currently bazel run does not attach the console to the process; all the
output is currently piped through bazel, and input is ignored (as you
noticed).

How about using the client/server mode?
http://lldb.llvm.org/remote.html

**Comment by helenalt** [comment]
Closing due to no activity from the requester for 2 years. If this feature is still desired, please submit a proposal to bazel-discuss@googlegroups.com

**Comment by laszlocsomor** [comment]
Didn't @lberki recently do something in this domain in https://github.com/bazelbuild/bazel/commit/4741aa83145fd2d3ab0599314f7732c5b80977bd ?

**Comment by lberki** [comment]
Yep, that sounds like a viable approach. After that commit, for example, `bazel run --run_under=gdb` will launch an interactive `gdb` session with a C++ test.

**Comment by ulfjack** [comment]
Indeed, please try it with a Bazel binary after the commit mentioned above - we're now attaching the binary run directly to the terminal.

**Comment by lberki** [comment]
...if you add the `--direct_run` command line option to `bazel run`. The plan is to make that the default, but that will take some time; I'm hoping for ~two weeks, but since I also want to flip the flag internally at Google, I'm somewhat worried that it breaks some weird use cases so there may be delays.
